### ISSUE TITLE
Global namespace pollution and possible memory leaks

### ISSUE THREAD

**Issue by Maus3rVonDutch** [angular-ui/ui-grid#4952]
Hi,

I found some serious problems in the current implementation of some directives concerning polluting the global namespace (i.e. the *Window* object). The problem occurs when using the `'var self = this;'` construction in the linking function of a directive. An example occurs in the *uiGridMenu* directive:
```
.directive('uiGridMenu', ...
  var uiGridMenu = {
    ...
    link: function ($scope, $elm, $attrs, uiGridCtrl) {
      var self = this;
```	 
	 
This seems like a good implementation of the *Method Invocation Pattern*, normally the `this` parameter would be a reference to the object on which the function is defined. However the linking function here is **not** a *Method Invocation Pattern*, but a *Apply Invocation Pattern* instead. To verify this, just set a breakpoint at that line number and inspect the previous (anonymous) function call. You will notice the function is invoked by angular as:
```
 function cloneAndAnnotateFn(fn, annotation) {
      return extend(function() { return fn.apply(null, arguments); }, fn, annotation);
    }
```	
The *Apply Invocation Pattern* lets you choose which object should be passed as the `this` parameter, which in this case is actually `null`. However because this specific directive is currently **not running in strict mode** (i.e. *'use strict'*) the following applies (see [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply)) documentation :
```
if the method is a function in non-strict mode code, null and undefined will be replaced with the global object, and primitive values will be boxed
```
The `this` parameter in other directives is also replace by the global *Window* object (for instance in the *uiGridColumnMenu* directive). At first glance this is not a *Apply Invocation Pattern*, but I did not inspect it thoroughly, so I'm not really sure what the problem is for that particular directive. In any case, I think all directives need to be reviewed thoroughly and I'd recommend the following:

- all code should be run in strict mode (so you are warned early when such problems occur)
- all directives using the `var self = this;` in linking functions should be rewritten. 

For the latter, you have a few options in how to do this. For a quick fix (in the example of the 'uiGridMenu' directive), you could replace the line 

```
'var self = this;' 
```
with 
```
'var self = uiGridMenu;'
```
However I would recommend a different approach using a directive controller instead. This is the better option because of a number of things like testing, reuseability etc, but also because you do not have the problem concerning the `this` parameter, it will be bound to the object as you expect it to be.

I do not know whom to address this issue to, but I'm hoping any members of the core team will pick up on this issue.

**Comment by swalters** [comment]
@Maus3rVonDutch thanks for detailed issue.  I was able to confirm the issue and will make the quick fixes.

**Comment by swalters** [comment]
I can only find this issue on ui-grid-column-menu directive.  @Maus3rVonDutch do you have specific examples of other directives doing this?  A quick search turns up nothing for me.

ui-grid-menu only has:
` controller: ['$scope', '$element', '$attrs', function ($scope, $element, $attrs) {
      var self = this;
    }]
'''
which doesn't do anything.

**Comment by Maus3rVonDutch** [comment]
@swalters hey and thnx for the quick response! 
I'm sorry, I forgot to mention I only reviewed the release file of version [3.0.7] (https://github.com/angular-ui/ui-grid.info/blob/gh-pages/release/3.0.7/ui-grid.js) and not the individual source files. On line 2000 you will see the following:
```
 templateUrl: 'ui-grid/uiGridMenu',
    replace: false,
    link: function ($scope, $elm, $attrs, uiGridCtrl) {
      var self = this;
      var menuMid;
      var $animate;

      $scope.i18n = {
        close: i18nService.getSafeText('columnMenu.close')
      };
```

I do not see that code in the [current source file](https://github.com/angular-ui/ui-grid/blob/master/src/js/core/directives/ui-grid-menu.js), so I guess it was fixed already by some other implementation. That file (and others) still lacks the `'use strict'` option though, which I really would recommend to prevent other not obvious bugs.

Here are the other occurrences in version 3.0.7:
* uiGridColumnMenu: 576 (but you fixed that already)
* uiGridPinnedContainer: 3452 (not sure if that's causing a problem because it is in a compile function and not in a linking function, I did not test it myself)
